staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go

```diff
@@ -542,6 +547,40 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
 
 		info.Refresh(obj, true)
 
+		// Migrate managed fields if necessary.
+		//
+		// By checking afterward instead of fetching the object beforehand and
+		// unconditionally fetching we can make 3 network requests in the rare
+		// case of migration and 1 request if migration is unnecessary.
+		//
+		// To check beforehand means 2 requests for most operations, and 3
+		// requests in worst case.
+		if err = o.saveLastApplyAnnotationIfNecessary(helper, info); err != nil {
+			fmt.Fprintf(o.ErrOut, warningMigrationLastAppliedFailed, err.Error())
+		} else if performedMigration, err := o.migrateToSSAIfNecessary(helper, info); err != nil {
```

Comment:
Should we do the `saveLastApplyAnnotationIfNecessary` if `performedMigration` is false?

Reply:
I dont think you should save the annotation after migrating. Ordering it that way, if preserving the annotation fails then next time user invokes server-side-apply the annotation would be dropped.